tools/integration_tests/list_large_dir/list_dir_with_twelve_thousand_files_test.go

```diff
@@ -99,14 +104,75 @@ func checkIfObjNameIsCorrect(t *testing.T, objName string, prefix string, maxNum
 	}
 }
 
+func splitBucketNameAndDirPath(bucketNameWithDirPath string, t *testing.T) (bucketName, dirPathInBucket string) {
+	t.Helper()
+
+	idx := strings.Index(bucketNameWithDirPath, "/")
+	if idx <= 0 {
+		t.Errorf("Unexpected bucketNameWithDirPath: %q. Expected form: <bucket>/<object-name>", bucketNameWithDirPath)
+	}
+	bucketName = bucketNameWithDirPath[:idx]
+	dirPathInBucket = bucketNameWithDirPath[idx+1:]
+	return
+}
+
+func testdataUploadFilesToBucket(ctx context.Context, storageClient *storage.Client, bucketNameWithDirPath, dirWithTwelveThousandFiles, filesPrefix string, t *testing.T) {
+	t.Helper()
+
+	bucketName, dirPathInBucket := splitBucketNameAndDirPath(bucketNameWithDirPath, t)
+
+	dirWithTwelveThousandFilesFullPathPrefix := filepath.Join(dirWithTwelveThousandFiles, filesPrefix)
+	matches, err := filepath.Glob(dirWithTwelveThousandFilesFullPathPrefix + "*")
+	if err != nil {
+		t.Fatalf("Failed to get files of pattern %s*: %v", dirWithTwelveThousandFilesFullPathPrefix, err)
+	}
+
+	type copyRequest struct {
+		srcLocalFilePath string
+		dstGcsObjectPath string
+	}
+	numCopyGoroutines := 16
+	copyGoroutines := make([][]copyRequest, numCopyGoroutines)
+	copyRequestCounter := 0
+	for i := 0; i < numCopyGoroutines; i++ {
+		copyGoroutines[i] = []copyRequest{}
+	}
+
+	for _, match := range matches {
+		_, fileName := filepath.Split(match)
+		if len(fileName) > 0 {
+			copyGoroutineIndex := copyRequestCounter % numCopyGoroutines
+			copyGoroutines[copyGoroutineIndex] = append(copyGoroutines[copyGoroutineIndex], copyRequest{srcLocalFilePath: match, dstGcsObjectPath: filepath.Join(dirPathInBucket, fileName)})
+			copyRequestCounter++
+		}
+	}
+
+	wg := sync.WaitGroup{}
+	for _, copyGoroutine := range copyGoroutines {
+		wg.Add(1)
+		go func() {
+			for _, copyRequest := range copyGoroutine {
+				client.CopyFileInBucket(ctx, storageClient, copyRequest.srcLocalFilePath, copyRequest.dstGcsObjectPath, bucketName)
+			}
+			wg.Done()
+		}()
+	}
+	wg.Wait()
```

Comment:
Use channels to limit concurrency.
```suggestion
	numCopyGoroutines := 16
	ch := make(chan copyRequest, numCopyGoroutines)
	go func() {
	  for _, match := range matches {
	    if len(fileName) == 0 {
	      continue
	    }
	    ch <- copyRequest{
	        srcLocalFilePath: match,
	        dstGcsObjectPath: filepath.Join(dirPathInBucket, fileName),
	      }
	  }
	  close(ch)
	}()
	for {
	  select {
	    case cpReq, ok := <- ch:
	      if !ok {
	        break
	      }
	      client.CopyFileInBucket(ctx, storageClient, cpReq.srcLocalFilePath, cpReq.dstGcsObjectPath, bucketName)	      
	  }
	}
```

Reply:
Done. Switched to a single unbuffered channel as discussed offline.

Stats associated with this change:

time taken to copy all  1000   files, using arrays in goroutines                               =    7.929863456s
time taken to copy all  1000   files, using arrays in goroutines                               =    6.787053469s
time taken to copy all  1000   files, using arrays in goroutines                               =    6.375443899s
time taken to copy all  1000   files, using a single buffered channel across all goroutines    =    7.851011189s
time taken to copy all  1000   files, using a single buffered channel across all goroutines    =    6.51152162s
time taken to copy all  1000   files, using a single buffered channel across all goroutines    =    6.14107131s
time taken to copy all  1000   files, using a single unbuffered channel across all goroutines  =    7.013791204s
time taken to copy all  1000   files, using a single unbuffered channel across all goroutines  =    6.269052254s
time taken to copy all  1000   files, using a single unbuffered channel across all goroutines  =    6.056092328s
time taken to copy all  12000  files, using a single unbuffered channel across all goroutines  =  1m7.534551842s
time taken to copy all  12000  files, using a single unbuffered channel across all goroutines  =  1m5.070493206s
time taken to copy all  12000  filesi, using a single unbuffered channel across all goroutines =  1m4.751596961s